### ISSUE TITLE
SignalR tolerance for packet loss

### ISSUE THREAD

**Issue by RossMetacraft** [dotnet/aspnetcore#54725]
Greetings all,

This is a bit of a long post to describe the issue, but I want to give as much context as I can.

TL;DR:

I'm trying to send frequent (200 per second) small (~500 byte) messages to SignalR clients, and if there is packet loss on a given client's connection, these messages arrive late, and the lag accumulates over time, only for that one client. Other clients do not experience this lag as long as they don't have any packet loss on their own connection.

Full context:

I have a SignalR application which is essentially a multiplayer game server. The server is self-hosted as a single server with no scale-out. Clients use only websockets, not any of the other transports.

This application is for a flight simulation community where the clients are WPF desktop apps which simulate air traffic control radar, and the server sends each client information about where each aircraft is located, once per second. During busy periods, there may be around 200 aircraft connected to the server, so each client is receiving 200 aircraft position updates per second. These updates are sent to clients by calling a client method which has a `List<AircraftUpdate>` as the parameter. (In other words, each client is having its `ReceiveAircraftUpdates` method called once per second, with a list of 200 `AircraftUpdates` passed in as the single parameter. Each update is around 500 bytes of JSON.)

Clients also call server methods when the user takes an action in the WPF client, such as updating information about an aircraft's flight plan.

This all works very well 99% of the time. However, if a client has a poor connection to the server, and there is packet loss occurring anywhere along the route, that's when we start to see problems. Latency is not a problem as long as it's relatively consistent ... it's packet loss that is the issue. The symptoms are that the aircraft position updates will start to "fall behind". The client with the poor connection will see the aircraft on their radar scope in a position that is some number of seconds behind where the actual aircraft is located. So it's like time is slowing down for them. Everything continues working properly for all other clients that have a stable connection.

On the client with the poor connection, the aircraft positions will fall further and further behind reality as time goes on. If the packet loss is resolved, the issue quickly resolves itself and the aircraft appear to speed up to catch up with reality. If the packet loss returns, the lag starts to accumulate again.

If a client is currently lagging in this way, and they take an action (like updating an aircraft's flight plan) that results in calling a server method, that method is called immediately on the server, but the response (the return value of the server method) is not received by the client until after an amount of lag that matches the lag currently being experienced in the aircraft position updates.

Not surprisingly, if the client disconnects and reconnects, it resets the lag back to zero and the lag will then begin accumulating again, assuming the connection still has packet loss.

I made simple server+client repro applications and I'm able to reproduce these conditions reliably. I host the repro server on a simple Linux VM, and the repro client is a WPF app running on my desktop. I introduce artificial packet loss using the "Clumsy" app. I configure Clumsy to introduce a 10% chance of packet loss on the https connection to the repro server. The server is continually calling a `ReceiveData` method on the client, passing a list of 200 strings, each string is about 500 bytes in length. This roughly simulates position update data for 200 aircraft being sent every second. The `ReceiveData` client method also takes an integer parameter, and the server increments this integer each time it calls the client method, so that I can easily see how far behind the client is getting. If I enable the 10% packet loss in Clumsy, I immediately start to see the client receiving the data after some amount of lag, and the amount of lag grows over time, at a rate of about half a second per second. So after 30 seconds, the client is receiving the data about 15 seconds behind. If I disable Clumsy, the client then quickly catches up over the next few seconds, and then receives the data instantly as normal. Obviously, if I alter the packet loss % that Clumsy is introducing, the rate of lag accumulation will be higher or lower accordingly.

My assumption is that these issues are simply due to websockets using TCP, which guarantees ordered delivery, and SignalR is simply buffering any messages sent to the client (including calls to client methods as well as return values of server methods called by the client) in a FIFO queue, and this buffer is just filling up faster than it can be emptied due to the 10% packet loss. When the packet loss goes away, this buffer is able to be emptied and things return to normal.

My understanding is that SignalR is used for real time web-based games, and I have to imagine that these games need to be able to tolerate connections with packet loss, without the state of game objects on the clients falling further and further behind that of the server. So I'm left wondering if I'm doing something wrong and there is a better way to send these aircraft position updates to the client.

I'm considering using UDP datagrams to send the aircraft position updates, since I don't really care if some of these updates are never received by the client. However, I would obviously much prefer to keep everything over the SignalR connection if possible.

I should also note that I tried using SignalR streaming methods to send the aircraft position updates, but this had no effect. (I wasn't really expecting it to, but I gave it a shot anyway.)

In short, is SignalR just not the right tool for the job when clients may have some packet loss? Is there any way to tune SignalR to drop certain messages from its outgoing buffer if the buffer gets too full?

**Comment by davidfowl** [comment]
This well thought out post reminded me of [this issue](https://github.com/dotnet/aspnetcore/issues/43093). I wonder if the techniques from there can help you in this situation.

**Comment by RossMetacraft** [comment]
> This well thought out post reminded me of [this issue](https://github.com/dotnet/aspnetcore/issues/43093). I wonder if the techniques from there can help you in this situation.

Thanks very much for the quick reply, David. I read through that other issue and studied your commits. It does seem like we are facing similar issues and we're both looking for a similar solution, that being a way to have these outgoing packets dropped when there is backpressure.

I think the two techniques discussed in that other thread that might be helpful in my situation are the use of a bounded channel for each client (so that any backpressure will cause older packets to be discarded) or the use of a cancellation token with a timeout.

I think I have a good grasp of how the bounded channel would work, however, I'm not sure I fully understand how the cancellation token would work. It seems like the token method would be simpler to implement (doesn't require doing my own buffering with bounded channels) so I would like to explore that more. From your comments (both in the GH thread and in your code) I can't tell if the token cancellation would cause the client to disconnect, or if it would simply cancel the single message sending attempt. (Essentially pulling that message out of the client's send queue, thereby relieving some backpressure.) Would you mind elaborating a bit on that technique?

**Comment by davidfowl** [comment]
>  From your comments (both in the GH thread and in your code) I can't tell if the token cancellation would cause the client to disconnect, or if it would simply cancel the single message sending attempt.

It would cancel the send attempt. I never spent time implementing this but it conceptually there's really only one solution. While there's an outbound incomplete send to a client, don't enqueue any items to that client. Cancellation just lets you stop the send but doesn't stop the additional queueing that might be happening in the meantime.

Perhaps an even more efficient solution would be use a single buffer for the event that all clients pull from. This would be similar to what .NET Framework SignalR was able to do. 

Stick with the channel per client implementation, it's much simpler, efficient enough and easier to understand.

**Comment by RossMetacraft** [comment]
Makes sense ... thanks very much for the input!

**Comment by davidfowl** [comment]
No, I think we should wait for an ack from @RossMetacraft once he's implemented something.

**Comment by RossMetacraft** [comment]
I experimented a bit with the channel-per-client technique, and I think I could get it to work reasonably well for my application after a good amount of testing and tuning. However, I decided that what I'm really looking for here (in concept) is unreliable fire-and-forget message sending, and I feel like there's already a battle-tested solution for that, namely UDP packets. So I'm adding a UDP connection along side the SignalR connection.

Just out of curiosity, has the team considered building in support for prioritizing calls in some way? E.g. there could be a way for the application code to tell SignalR that a given client method call is low-priority and can be discarded (or bypassed in favor of higher-priority calls in the queue) if there is any backpressure on the client connection.

Thanks again for the help!

**Comment by davidfowl** [comment]
> Just out of curiosity, has the team considered building in support for prioritizing calls in some way? E.g. there could be a way for the application code to tell SignalR that a given client method call is low-priority and can be discarded (or bypassed in favor of higher-priority calls in the queue) if there is any backpressure on the client connection.

Not seriously, no. Making the protocol this much more complex and implementing it on clients and server would require a large number of people running into these scenarios enough that it would make sense to invest in. That hasn't happened yet and the workaround is pretty good (IMO). 

I do question whether we should have something build in for this though like the HubManager code in the linked issue.

**Comment by RossMetacraft** [comment]
Understood, make sense. I still find it odd that this issue of slowly-increasing backpressure for connections with moderate packet loss hasn't come up more often. I was fully expecting to find more relevant discussions when I searched around initially. When I found very little, I thought sure I was just doing something wrong in my application, until I made the very basic repro client+server and found the same symptoms there.

Is SignalR just not being used as much as I thought for things like real time games with a lot of entity position updates getting pushed to clients to keep them in sync with the game state on the server?

**Comment by davidfowl** [comment]
I think it is but its possible people controlling it by sending to individual clients and doing manual grouping.

**Comment by RossMetacraft** [comment]
> I think it is but its possible people controlling it by sending to individual clients and doing manual grouping.

By "controlling it" do you mean limiting the negative impact to just the client(s) with a poor connection? If so, that's already the case for me. If client A has bad packet loss it doesn't affect client B, even if they are both being called at the same time by the same client method invocation. I suspect this is because I treat these client methods as fire-and-forget, by not awaiting the call.

So it's good that the lag issues are limited to just the client(s) with a bad connection, but this still leaves them with a very poor experience, of course. And it does ultimately negatively impact the clients with a good connection, for example when client A (good connection) tries to transfer control of an aircraft to client B (poor connection), client B won't see the transfer request for some time, because of the backpressure-induced lag ... the transfer request message will sit in their outgoing queue for a while. So user A thinks that user B is just ignoring the transfer request, so they send them a text message, and of course that lags as well.

**Comment by davidfowl** [comment]
> By "controlling it" do you mean limiting the negative impact to just the client(s) with a poor connection? If so, that's already the case for me. If client A has bad packet loss it doesn't affect client B, even if they are both being called at the same time by the same client method invocation. I suspect this is because I treat these client methods as fire-and-forget, by not awaiting the call.

Yes, if you are sending to individual clients, then you can control what happens if that client is slow, especially now that we have the ability for observe back pressure outside of just the TCP buffer (but using ACKs from the client).

> the transfer request message will sit in their outgoing queue for a while. So user A thinks that user B is just ignoring the transfer request, so they send them a text message, and of course that lags as well.

That's where the combination of cancellation and the bounded channel come in. Network latency is what it is (physics), the difference is what behaviors you apply when backpressure is observed. If a client never reads the message (consider an extreme case) the nothing will save you from the delay.

**Comment by RossMetacraft** [comment]
> especially now that we have the ability for observe back pressure outside of just the TCP buffer (but using ACKs from the client).

Here you are referring to the ability to have a return value from a client method call, right? So if I wanted to go down this road, I would implement a solution that uses a bounded channel and only adds items to the channel if there are no previous calls still in flight. And I would put a short (maybe around 1 second) cancellation time on each send as well. Do I have that right?

**Comment by davidfowl** [comment]
That's right.

**Comment by RossMetacraft** [comment]
Okay, thanks. I'm already just about done with the UDP solution, so I'm going to stick with that for now, but I may revisit this option in the future. It definitely would be nice to keep everything on SignalR if possible.

Thanks very much for the discussion ... very helpful and appreciated!

**Comment by davidfowl** [comment]
Thanks for the great discussion as well!
